File: iosApp/iosApp/SwiftUIDemo/DemoSession.swift
import Foundation
import ComposeApp

final class DemoSession: ObservableObject {
    static let shared = DemoSession()

    let manager = AudioSessionManager()
    @Published var interrupted = false
    @Published var routeLabel: String?

    private var interruptionListener: ((AudioInterruptionInfo) -> Void)?

    private init() {
        do {
            try manager.configure(category: .playAndRecord, mode: .default_)
            try manager.setActive(active: true)
        } catch {
            NSLog("KlarinetDemo session configure failed: \(error)")
        }
        let listener: (AudioInterruptionInfo) -> Void = { [weak self] info in
            DispatchQueue.main.async {
                self?.interrupted = info.type == .began
            }
        }
        interruptionListener = listener
        manager.observeInterruptions(listener: listener)
        manager.observeRouteChanges { [weak self] info in
            DispatchQueue.main.async {
                self?.routeLabel = "Route: \(info.reason)"
            }
        }
    }

    func requestRecordPermission(_ onResult: @escaping (Bool) -> Void) {
        manager.requestRecordPermission { granted in
            DispatchQueue.main.async { onResult(granted.boolValue) }
        }
    }

    func attach(_ stream: AudioStream) {
        manager.attach(stream: stream)
    }

    func detach(_ stream: AudioStream) {
        manager.detach(stream: stream)
    }
}

File: demo/src/commonMain/kotlin/com/vectencia/klarinet/demo/DemoSession.kt
package com.vectencia.klarinet.demo

import com.vectencia.klarinet.AudioInterruptionInfo
import com.vectencia.klarinet.AudioInterruptionType
import com.vectencia.klarinet.AudioRouteChangeInfo
import com.vectencia.klarinet.AudioSessionCategory
import com.vectencia.klarinet.AudioSessionManager
import com.vectencia.klarinet.AudioSessionMode
import com.vectencia.klarinet.AudioStream
import com.vectencia.klarinet.coroutines.interruptionFlow
import com.vectencia.klarinet.coroutines.routeChangeFlow
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

object DemoSession {
    val manager: AudioSessionManager = AudioSessionManager()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    private val _interruption = MutableStateFlow<AudioInterruptionInfo?>(null)
    val interruption: StateFlow<AudioInterruptionInfo?> = _interruption.asStateFlow()

    private val _routeChange = MutableStateFlow<AudioRouteChangeInfo?>(null)
    val routeChange: StateFlow<AudioRouteChangeInfo?> = _routeChange.asStateFlow()

    init {
        try {
            manager.configure(AudioSessionCategory.PLAY_AND_RECORD, AudioSessionMode.DEFAULT)
            manager.setActive(true)
        } catch (_: Exception) {
        }
        scope.launch {
            manager.interruptionFlow().collect { info ->
                _interruption.value = if (info.type == AudioInterruptionType.BEGAN) info else null
            }
        }
        scope.launch {
            manager.routeChangeFlow().collect { info ->
                _routeChange.value = info
            }
        }
    }

    fun hasRecordPermission(): Boolean = manager.hasRecordPermission()

    fun requestRecordPermission(onResult: (granted: Boolean) -> Unit) {
        manager.requestRecordPermission(onResult)
    }

    fun attach(stream: AudioStream) {
        manager.attach(stream)
    }

    fun detach(stream: AudioStream) {
        manager.detach(stream)
    }
}
